### ISSUE TITLE
Variable scoping order issue

### ISSUE THREAD

**Issue by ceymard** [satyr/coco#100]
Don't know if bug or not. The following doesn't compile ;

``` coffeescript
function a 

   function b
       v := 2

   v = null
   b!
```

since `v` is declared in `a` and is evaluated before `b` is called (since coco puts all the vars at the beginning of the scope), I think this should be allowed to compile.

IE. nested functions should be evaluated after their parent scope.

**Comment by satyr** [comment]
This is because variables are registered from top to bottom as they are compiled.
To fix, we'd have to scan all the variable declarations in each function before compiling it.

See also: https://github.com/jashkenas/coffee-script/issues/1500

**Comment by ceymard** [comment]
Yes this is what I thought ; which is why I was actually suggesting deferring the parsing of nested functions after parsing the toplevel one.

**Comment by ceymard** [comment]
Another "solution" (albeit not very pretty) would be to have the `:=` ignore variable definitions and just translate code to `v = 2;` without further inspection.

This actually would make sense for scripts meant to be included in the browser where a global variable may be defined in another script. I agree it's ugly, but there is a use case.

Another quick solution I can imagine would be parsing the whole tree and have the whole variable checking done _after_ we determine all the variables we have to output.

**Comment by satyr** [comment]
> Another "solution" (albeit not very pretty) would be to have the `:=` ignore variable definitions and just translate code to `v = 2;` without further inspection.

Can't do that. Assigning to undeclared variables should be forbidden as per ES5 strict.
